ipa-core/src/report/hybrid.rs

```diff
@@ -224,50 +385,135 @@ impl<BK, V> HybridReport<BK, V>
-    ) -> Result<Vec<u8>, InvalidReportError> {
-        unimplemented!()
+        key_id: KeyIdentifier,
+        key_registry: &impl PublicKeyRegistry,
+        info: &HybridInfo,
+        rng: &mut R,
+    ) -> Result<Vec<u8>, InvalidHybridReportError> {
+        match self {
+            HybridReport::Impression(impression_report) => match info {
+                HybridInfo::Impression(impression_info) =>
+                    impression_report.encrypt(key_id, key_registry, impression_info, rng).map(|v| once(HybridEventType::Impression as u8).chain(v).collect()),
+                HybridInfo::Conversion(_) => {
+                    Err(InvalidHybridReportError::WrongInfoType("Impression"))
+                }
+            },
+            HybridReport::Conversion(conversion_report) => match info {
+                HybridInfo::Conversion(conversion_info) =>
+                    conversion_report.encrypt(key_id, key_registry, conversion_info, rng).map(|v| once(HybridEventType::Conversion as u8).chain(v).collect()),
+                HybridInfo::Impression(_) => {
+                    Err(InvalidHybridReportError::WrongInfoType("Conversion"))
+                }
+            },
+        }
+    }
+
+    /// # Errors
+    /// If there is a problem encrypting the report.
+    pub fn encrypt_to<R: CryptoRng + RngCore, B: BufMut>(
+        &self,
+        key_id: KeyIdentifier,
+        key_registry: &impl PublicKeyRegistry,
+        info: &HybridInfo,
+        rng: &mut R,
+        out: &mut B,
+    ) -> Result<(), InvalidHybridReportError> {
+        match self {
+            HybridReport::Impression(impression_report) => match info {
+                HybridInfo::Impression(impression_info) =>{
+                    out.put_u8(HybridEventType::Impression as u8);
+                    impression_report.encrypt_to(key_id, key_registry, impression_info, rng, out)},
+                HybridInfo::Conversion(_) => {
+                    Err(InvalidHybridReportError::WrongInfoType("Impression"))
+                }
+            },
+            HybridReport::Conversion(conversion_report) => match info {
+                HybridInfo::Conversion(conversion_info) =>{
+                    out.put_u8(HybridEventType::Conversion as u8);
+                    conversion_report.encrypt_to(key_id, key_registry, conversion_info, rng, out)},
+                HybridInfo::Impression(_) => {
+                    Err(InvalidHybridReportError::WrongInfoType("Conversion"))
+                }
+            },
+        }
     }
 }
 
 /// `HybridImpressionReport`s are encrypted when they arrive to the helpers,
 /// which is represented here. A `EncryptedHybridImpressionReport` decrypts
 /// into a `HybridImpressionReport`.
-#[derive(Copy, Clone, Eq, PartialEq)]
-pub struct EncryptedHybridImpressionReport<BK, B>
+#[derive(Clone, Eq, PartialEq)]
+pub struct EncryptedHybridImpressionReport<BK>
 where
-    B: Deref<Target = [u8]>,
     BK: SharedValue,
 {
-    data: B,
+    data: Bytes,
```

Comment:
this feels much cleaner, but I just want to make sure @akoshelev sees this, since he may know the history of why we used `B: Deref<Target = [u8]>` previously, and can confirm we don't need it anymore.

Reply:
yea, we chatted about it with @tyurek. We need to inspect the first byte and cut it off before sending it down for further processing. It is not possible with the `B` bounds. `Bytes` would provide similar ergonomics